doc/mgmt/config_replace_rollback_HLD.md

```diff
@@ -0,0 +1,418 @@
+
+
+When a configure replace resquest is sent from non-configure session or REST, gNMI/gNOI:
+- Check if CONFIG_DB is locked. If it is locked, the command is rejected to avoid conflicting configuration state.
+
+When a configure replace request is sent within a configure session:
+- When 'replace' command is issued, check if the candidate store is dirtified. If it is dirtified, the command is rejected to avoid conflicting configuration state.
+- Since DB writes are targeted to candidate datastore, CLI 'replace' command is followed by commit with/wthout <timer> to apply changes to CONFIG_DB.
+- "commit" command without a timer saves the changes to CONFIG_DB. Chekcpoints will be generated by configure session.
+- "commit \<timer\>" command saves the changes to CONFIG_DB and put in trial stage until it is confirmed. A temporarty backup file for current config will be saved for rollback. if a timer gets expired without confirm, it will be rolled back to the previous configuration state by internal config replacement with saved backup file. If an error occurred during rollback, it may require manual troubleshooting to fix the problem state.
+- "commit confirm" command to confirm new configuration changes that are applied to running-configuration, and move from trial to permanent stage. Chekcpoints will be generated by configure session. 
+
+  
+### 1.3.2 Container
+No new containers are introduced for this feature. Existing mgmt-framework container is updated.
+
+### 1.3.3 SAI Overview
+No new or existing SAI services are required
+
+
+# 2 Functionality
+
+Following functionalities are provided to support Config Replace and Rollback.
+- timer goroutine started by a timer option
+- automatic revert to the saved backup configuration if a timer gets expired
+- revert back to previous configuration state if "kill config session" is executed
+- cancel a timer 
+- reject the operation if candidate datastore is dirtified
+- action to support both session and session-less
+- action to archive CONFIG_DB to the backup configuration for rollback
+- modeless rpc "commit" to support timer
+- modeless rpc "commit" to to support confirm
+- add a rule to rbac to allow only admin user
+- add support to retrieve configuration file from remote server and decompress the file if it is in compressed format
+- error reporting
+- debug utilities
+
+# 3 Design
+## 3.1 Overview
+A new config replacement componenent on the server will be a key implemenation to orchestrate config replace operation requested by different NBIs. It loads a desired config file specified by a request and compares against CONFIG_DB to generate the diff results tagged with operation, table, keys and fields. The resultant diff data set will be sorted by operation - DELETE, CREATE and REPLACE in this order - as we do for REST/gNMI operation, and subsequently sorted by table orders learnt from CVL to apply the changes as a single transaction to CVL/DB layer.
```

Comment:
Please elaborate on diff operation. Is it a simple json_diff OR the yang constraints like leafref etc are also considered in the diff step?

Why are the operations ordered by DELETE->CREATE and not CREATE->DELETE?